### PR TITLE
#158911629 Refactor code

### PR DESCRIPTION
**Pull request by valentineezeh**
### What does this PR do?
This PR is a implementation of refactor codes of the controllers.

### Description of the task to be completed
- refactor codes in controllers
- implemented more test for API endpoints

### How can this be manually tested
- API endpoints can be tested via postman

### Relevant pivotal Tracker
[158911629](https://www.pivotaltracker.com/story/show/158911629)

### CODE REVIEW

File: server/controller/rideRequest.js [hunk 1]
```diff
@@ -19,64 +16,62 @@ class RideRequestController{
...
             if(err){
-                return res.status(400).json({
+                const duplicateKeyError = "riderequests_userid_rideid_key";
+                if(err.message.search(duplicateKeyError) !== -1){
+                    return res.status(500).json({
+                        success: false,
+                        message: 'Error: You can only request for a ride once.'
+                    })
+                }
+                return res.status(404).json({
```
Comment (Ruqoyah) [inline server/controller/rideRequest.js]: This simply means if there is any error, the message the user is going to get ride does not exist even though ride exist and the error is something else.

File: server/controller/rides.js [hunk 2]
```diff
@@ -6,20 +6,16 @@ class RidesController{
...
         client.query('SELECT * FROM rides WHERE id = $1', [req.params.rideId], (err, data)=>{
             if (data){
-                //console.log(data.rows)
                 let newRide = [];
-                // let newRideNum = parseInt(newRide, 10);
-                // console.log(newRideNum)
                 data.rows.map((ride) => {
-                    //console.log(ride.id)
                     if(ride.id === parseInt(req.params.rideId, 10)){
                         newRide.push(ride)
```
Comment (Ruqoyah) [inline server/controller/rides.js]: Why are you still pushing? The ride would be in an array if you are doing that meanwhile it should be in object format

File: server/controller/rides.js [hunk 3]
```diff
@@ -6,20 +6,16 @@ class RidesController{
...
             res.status(200).json({
                 success: true,
-                message: 'All Rides has been retrieve.',
-                data: data.rows
+                message: 'All rides has been retrieved.',
+                rides: data.rows
             })
         })  
     }
     static getARide(req, res, next){
```
Comment (Ruqoyah) [inline server/controller/rides.js]: I don't understand why you had to do getARide like this. I think when you select all where the id is what you provided in the param, the data that is going to be in the row would be the one that has the id you queried with, why do you have to loop again to check if the id is the same as you passed? the check that is supposed to have happened here is if data.row is not empty return the result else, return no ride found and then you return err is there is err

File: server/controller/rideRequest.js [hunk 4]
```diff
@@ -5,11 +5,8 @@ class RideRequestController{
         client.query("SELECT * FROM riderequests WHERE  rideId =$1", [ req.params.rideId ], (err, data) => {
```
Comment (Ruqoyah) [inline server/controller/rideRequest.js]: The same feedback I gave for getARide is happening here too, kindly follow the feedback and fix this too.

File: server/controller/rides.js [hunk 5]
```diff
@@ -43,26 +38,23 @@ class RidesController{
...
+            req.body.stopPoint,
+            req.body.departureTime,
             req.decoded.userId
         ]
-        //console.log(req)
         client.query(text, values, (err, data) => {
            if(err) return res.status(500).json({
                success: false,
                message: 'Internal Server error',
                error: next(err)
```
Comment (Ruqoyah) [inline server/controller/rides.js]: why did you add next(err)? I think it's suppose to be err.message

File: server/controller/user.js [hunk 6]
```diff
@@ -8,14 +8,14 @@ dotenv.config();
...
             req.body.firstname, 
             req.body.lastname, 
             req.body.about, 
-            req.body.email, 
+            email, 
             req.body.password,
-            req.body.confirmPassword
         ]
         client.query(query, values, (err, data) => {
             if (err) return res.status(409).send({
```
Comment (Ruqoyah) [inline server/controller/user.js]: Nothing happens if there is error? and also why is the status code 409?

File: server/controller/user.js [hunk 7]
```diff
@@ -37,72 +37,48 @@ class UserController{
...
-                           message: 'Authentication failed, wrong Password'
-                       });
-                   }else{
-                       bcrypt.compare(req.body.password, dataPassword, (err, response) => {
-                           if(err){
-                               return res.status(500).json({
-                                   error: err.message
-                               })
-                           }else{
+            if(!(data.rows[0])){
```
Comment (Ruqoyah) [inline server/controller/user.js]: You could just do if(data.row.length < 1)

File: server/controller/user.js [hunk 8]
```diff
@@ -37,72 +37,48 @@ class UserController{
...
+                return res.status(404).json({
+                    success: false,
+                    message: 'User does not exist.'
+                })
+            }else if(data.rows[0]){
```
Comment (Ruqoyah) [inline server/controller/user.js]: then this would just be else {} instead of else if again

File: server/controller/user.js [hunk 9]
```diff
@@ -37,72 +37,48 @@ class UserController{
...
+                if(data.rows[0].password !== req.body.password){
```
Comment (Ruqoyah) [inline server/controller/user.js]: You don't need this check at all because the password would never be the same in any case. You are comparing what is not encrypted with what has not been encrypted, it is wrong.

File: server/controller/user.js [hunk 10]
```diff
@@ -37,72 +37,48 @@ class UserController{
...
-                       bcrypt.compare(req.body.password, dataPassword, (err, response) => {
-                           if(err){
-                               return res.status(500).json({
-                                   error: err.message
-                               })
-                           }else{
+            if(!(data.rows[0])){
+                return res.status(404).json({
+                    success: false,
+                    message: 'User does not exist.'
```
Comment (Ruqoyah) [inline server/controller/user.js]: This should be invalid credential and the status code is suppose to be 401 because you queried the table based on the email passed in by the user so if the email doesn't exist then it's an invalid credential

File: server/controller/user.js [hunk 11]
```diff
@@ -37,72 +37,48 @@ class UserController{
...
+                })
+            }else if(data.rows[0]){
+                if(data.rows[0].password !== req.body.password){
+                    res.status(400).json({
+                        success: false,
+                        message: 'Invalid Credentials'
+                    })
+                }else{
+                    bcrypt.compare(req.body.password, data.rows[0].password, (err, response) => {
```
Comment (Ruqoyah) [inline server/controller/user.js]: You already encrypting the password and comparing here to check if the password matches with the one in the database and if there is a response then it should sign up successfully but if there is an error, it should return invalid credentials.

File: server/controller/user.js [hunk 12]
```diff
@@ -37,72 +37,48 @@ class UserController{
...
+                        if(err){
+                            return res.status(500).json({
+                                success: false,
+                                message: err.message
```
Comment (Ruqoyah) [inline server/controller/user.js]: So, this should be message: "Invalid Credentials" and the status code should be 401 and not 500.

File: server/controller/user.js [hunk 13]
```diff
@@ -37,72 +37,48 @@ class UserController{
...
+                    success: false,
+                    message: 'User does not exist.'
+                })
+            }else if(data.rows[0]){
+                if(data.rows[0].password !== req.body.password){
+                    res.status(400).json({
+                        success: false,
+                        message: 'Invalid Credentials'
+                    })
+                }else{
```
Comment (Ruqoyah) [inline server/controller/user.js]: remove the else since there won't be any if statement again.

File: server/controller/user.js [hunk 14]
```diff
@@ -37,72 +37,48 @@ class UserController{
...
+                            return res.status(500).json({
+                                success: false,
+                                message: err.message
+                            })
+                        }else{
                             const token = jwt.sign({
-                                email: dataEmail,
-                                userId: dataUserId,
+                                userId: data.rows[0].id,
+                                email: data.rows[0].email,
```
Comment (Ruqoyah) [inline server/controller/user.js]: why did you add the user email to jwt????

File: server/middleware/auth.js [hunk 15]
```diff
@@ -11,18 +11,19 @@ const Auth = {
         if(token){
             jwt.verify(token, secret, (err, decoded) =>{
                 if(err){
-                    console.log(err);
                     return res.status(401).send({
                         message: 'You do not have permission to this page.'
                     });
                 }else{
                     // if everything is good, save to request for use in other routes
+                    //console.log(decoded)
```
Comment (Ruqoyah) [inline server/middleware/auth.js]: Take out console.log()

File: server/middleware/auth.js [hunk 16]
```diff
@@ -11,18 +11,19 @@ const Auth = {
...
                     req.decoded = decoded;
+                    //console.log(req.decoded.id)
```
Comment (Ruqoyah) [inline server/middleware/auth.js]: Take out console.log()

File: server/middleware/auth.js [hunk 17]
```diff
@@ -11,18 +11,19 @@ const Auth = {
...
                     next();
                 }
             });
         }else{
-            return res.status(401).send({ message: 'No token provided' })
+            return res.status(401).send({ message: 'User has either not Sign Up or Login.' })
```
Comment (Ruqoyah) [inline server/middleware/auth.js]: hmmm. don't think this error message is nice. It could be "You need to signup or login to perform this action"

File: server/middleware/postRideRequestValid.js [hunk 18]
```diff
@@ -1,24 +1,14 @@
...
+        if (req.body.status === undefined || req.body.status.toString().trim() === '') {
             return res.status(400).send({
-                message: 'Accept should not be empty.'
+                message: 'status should not be empty.'
               });
         }
-        if (req.body.accept !== "true" && req.body.accept !== "false") {
-            console.log(req.body.accept)
+        if (req.body.status !== "accept" && req.body.status !== 'reject') {
+            console.log(req.body.status)
```
Comment (Ruqoyah) [inline server/middleware/postRideRequestValid.js]: take out console.log()

File: server/middleware/postRideValidate.js [hunk 19]
```diff
@@ -1,37 +1,37 @@
 class PostValidation {
     static rideValidator(req, res, next) {
         //console.log('START')
```
Comment (Ruqoyah) [inline server/middleware/postRideValidate.js]: take out console.log()

File: server/middleware/postRideValidate.js [hunk 20]
```diff
@@ -1,37 +1,37 @@
...
             });
         }
         
-        if (req.body.destinationstartpoint.toString().trim() === '') {
+        if (req.body.startPoint.toString().trim() === '') {
             return res.status(400).send({
                 message: 'destination start point is required.'
             })
         }
         //console.log('========================================')
```
Comment (Ruqoyah) [inline server/middleware/postRideValidate.js]: take out console.log()

File: server/middleware/postRideValidate.js [hunk 21]
```diff
@@ -1,37 +1,37 @@
...
         //console.log('STOP')
```
Comment (Ruqoyah) [inline server/middleware/postRideValidate.js]: Take out console.log()

File: server/db/ride.sql [hunk 22]
```diff
@@ -1,14 +1,14 @@
 CREATE TABLE rides (
     id SERIAL PRIMARY KEY,
-    destinationStartPoint VARCHAR NOT NULL,
-    destinationStopPoint VARCHAR NOT NULL,
+    startPoint VARCHAR NOT NULL,
+    stopPoint VARCHAR NOT NULL,
     departureTime VARCHAR NOT NULL,
```
Comment (Ruqoyah) [inline server/db/ride.sql]: You should also have the departureDate too, I remember they gave you the feedback.

**Review by Ruqoyah — CHANGES_REQUESTED** [review]

File: server/controller/rideRequest.js (hunk 1 above)
Comment (valentineezeh) [reply server/controller/rideRequest.js]: please i don't understand what you mean here, can you please explain further?

File: server/controller/rides.js (hunk 2 above)
Comment (valentineezeh) [reply server/controller/rides.js]: Alright i have fix this.

File: server/controller/rides.js (hunk 3 above)
Comment (valentineezeh) [reply server/controller/rides.js]: I have fix this as well.

File: server/middleware/postRideValidate.js (hunk 21 above)
Comment (valentineezeh) [reply server/middleware/postRideValidate.js]: console.log has been removed

File: server/controller/rideRequest.js (hunk 4 above)
Comment (valentineezeh) [reply server/controller/rideRequest.js]: fixed.

File: server/controller/rides.js (hunk 5 above)
Comment (valentineezeh) [reply server/controller/rides.js]: fixed.

File: server/middleware/postRideRequestValid.js (hunk 18 above)
Comment (valentineezeh) [reply server/middleware/postRideRequestValid.js]: fixed.

File: server/middleware/postRideValidate.js (hunk 19 above)
Comment (valentineezeh) [reply server/middleware/postRideValidate.js]: fixed.

File: server/middleware/postRideValidate.js (hunk 20 above)
Comment (valentineezeh) [reply server/middleware/postRideValidate.js]: fixed.

File: server/db/ride.sql (hunk 22 above)
Comment (valentineezeh) [reply server/db/ride.sql]: fixed

File: server/controller/user.js (hunk 6 above)
Comment (valentineezeh) [reply server/controller/user.js]: 409 is for conflict if user with the same email details already exist.

File: server/controller/user.js (hunk 7 above)
Comment (valentineezeh) [reply server/controller/user.js]: fixed.

File: server/controller/user.js (hunk 8 above)
Comment (valentineezeh) [reply server/controller/user.js]: fixed.

File: server/controller/user.js (hunk 9 above)
Comment (valentineezeh) [reply server/controller/user.js]: fixed.

File: server/controller/user.js (hunk 10 above)
Comment (valentineezeh) [reply server/controller/user.js]: fixed.

File: server/controller/user.js (hunk 11 above)
Comment (valentineezeh) [reply server/controller/user.js]: okay its fixed.

File: server/controller/user.js (hunk 12 above)
Comment (valentineezeh) [reply server/controller/user.js]: fixed

File: server/controller/user.js (hunk 13 above)
Comment (valentineezeh) [reply server/controller/user.js]: fixed

File: server/controller/user.js (hunk 14 above)
Comment (valentineezeh) [reply server/controller/user.js]: it has been removed.

File: server/middleware/auth.js (hunk 15 above)
Comment (valentineezeh) [reply server/middleware/auth.js]: it has been removed.

File: server/middleware/auth.js (hunk 16 above)
Comment (valentineezeh) [reply server/middleware/auth.js]: it has been removed.

File: server/middleware/auth.js (hunk 17 above)
Comment (valentineezeh) [reply server/middleware/auth.js]: okay i have fixed that.

File: server/controller/user.js (hunk 6 above)
Comment (Ruqoyah) [reply server/controller/user.js]: But we would also get 409 for every err here because you don't just check if the email address exists, you checked if there is err return 409 which is wrong

File: server/controller/rideRequest.js (hunk 1 above)
Comment (Ruqoyah) [reply server/controller/rideRequest.js]: Okay, what I mean here is that you just checked if there is an err, return 404 user not found. it is wrong because we would always get user not found whenever there is an error regardless of what the error might be.

Comment (Ruqoyah) [reply server/controller/rideRequest.js]: Okay, what I mean here is that you just checked if there is an err, return 404 user not found. it is wrong because we would always get user not found whenever there is an error regardless of what the error might be.

File: server/controller/user.js (hunk 6 above)
Comment (Ruqoyah) [reply server/controller/user.js]: But we would also get 409 for every err here because you don't just check if the email address exists, you checked if there is err return 409 which is wrong.

File: server/controller/rides.js [hunk 23]
```diff
@@ -3,69 +3,56 @@ import client from '../db/index';
 class RidesController{
     static getAllRides(req, res, next){
         client.query("select * from rides", (err, data) => {
-            if (err) return next(err);
+            if (err) return next(err.message);
```
Comment (Ruqoyah) [inline server/controller/rides.js]: why do we still have next here and not returning the error to the user the way you have been returning it?

File: server/controller/user.js [hunk 24]
```diff
@@ -8,23 +8,23 @@ dotenv.config();
...
+            email, 
             req.body.password,
-            req.body.confirmPassword
         ]
         client.query(query, values, (err, data) => {
-            if (err) return res.status(409).send({
+            if (err) return res.status(500).send({
                     success: false,
-                    message: "User already exist."
+                    message: "User Already exist."
```
Comment (Ruqoyah) [inline server/controller/user.js]: If there is any error now, the error message the user would be getting is user already exist, irrespective of what the error is. I felt like the 500 is server error, if you want to validate of user already exist then you should do a check to do that, and the status code for that would now be 409.

File: server/controller/user.js [hunk 25]
```diff
@@ -37,72 +37,39 @@ class UserController{
...
-                           message: 'Authentication failed, wrong Password'
-                       });
-                   }else{
-                       bcrypt.compare(req.body.password, dataPassword, (err, response) => {
-                           if(err){
-                               return res.status(500).json({
-                                   error: err.message
-                               })
-                           }else{
+            if(data.rows[0] < 1){
```
Comment (Ruqoyah) [inline server/controller/user.js]: This is a hack. it should be  `if(data.rowCount === 0){`
